src/Microsoft.DotNet.Build.CloudTestTasks/DownloadFromAzure.cs

```diff
@@ -48,96 +50,64 @@ public override bool Execute()
-                                AccountKey,
-                                "GET",
-                                dateTime,
-                                request));
-                        return request;
-                    };
-
-                    XmlDocument responseFile;
-                    string nextMarker = string.Empty;
-                    using (HttpResponseMessage response = await AzureHelper.RequestWithRetry(Log, client, createRequest))
-                    {
-                        responseFile = new XmlDocument();
-                        responseFile.LoadXml(await response.Content.ReadAsStringAsync());
-                        XmlNodeList elemList = responseFile.GetElementsByTagName("Name");
-
-                        blobsNames.AddRange(elemList.Cast<XmlNode>()
-                                                    .Select(x => x.InnerText)
-                                                    .ToList());
-                        nextMarker = responseFile.GetElementsByTagName("NextMarker").Cast<XmlNode>().FirstOrDefault()?.InnerText;
-                    }
-                    while (!string.IsNullOrEmpty(nextMarker))
+                        AccountName = AccountName,
+                        AccountKey = AccountKey,
+                        ContainerName = ContainerName,
+                        FilterBlobNames = BlobNamePrefix,
+                        BuildEngine = this.BuildEngine,
+                        HostObject = this.HostObject
+                    };                    
+                    listAzureBlobs.Execute();
+                    blobNames = listAzureBlobs.BlobNames.ToList();
+                }
+                else
+                {
+                    blobNames = BlobNames.Select(b => b.ItemSpec).ToList<string>();
+                    if (BlobNamePrefix != null)
                     {
-                        urlListBlobs = string.Format($"https://{AccountName}.blob.core.windows.net/{ContainerName}?restype=container&comp=list&marker={nextMarker}");
-                        using (HttpResponseMessage response = AzureHelper.RequestWithRetry(Log, client, createRequest).GetAwaiter().GetResult())
-                        {
-                            responseFile = new XmlDocument();
-                            responseFile.LoadXml(response.Content.ReadAsStringAsync().GetAwaiter().GetResult());
-                            XmlNodeList elemList = responseFile.GetElementsByTagName("Name");
-
-                            blobsNames.AddRange(elemList.Cast<XmlNode>()
-                                                        .Select(x => x.InnerText)
-                                                        .ToList());
-
-                            nextMarker = responseFile.GetElementsByTagName("NextMarker").Cast<XmlNode>().FirstOrDefault()?.InnerText;
-                        }
+                        blobNames = blobNames.Where(b => b.StartsWith(BlobNamePrefix)).ToList<string>();
```

Comment:
The performance hit comes from the fact that you are downloading a list of all the blob names which will become huge overtime. The storage APIs are optimized based on blob name prefix which is why if we can use a common prefix we won't need to download all the names but instead just the subset under the prefix.